generation/src/generation.rs

```diff
@@ -427,6 +536,135 @@ impl Register {
         }
         .into_token_stream()
     }
+
+    fn generate_block_definition(
+        &self,
+        registers: &RegisterCollection,
+        device: &Device,
+    ) -> TokenStream {
+        let cfg_attributes = &self.cfg_attributes;
+
+        let pascal_case_name_string = self.name.to_case(Case::Pascal);
+        let pascal_case_name = syn::Ident::new(&pascal_case_name_string, Span::call_site());
+        let snake_case_name = syn::Ident::new(&self.name.to_case(Case::Snake), Span::call_site());
+
+        let module_doc_string = format!("Implementation of the [{pascal_case_name}]({snake_case_name}::{pascal_case_name}) register block");
+
+        let doc_attribute = if let Some(description) = self.description.as_ref() {
+            quote! { #[doc = #description] }
+        } else {
+            quote!()
+        };
+
+        let address_type = if let Some(register_address_type) = device.register_address_type {
+            register_address_type.into_type()
+        } else {
+            return syn::Error::new(Span::call_site(), "register_address_type is not specified")
+                .to_compile_error();
+        };
+
+        let fns = match self.generate_block_functions(registers) {
+            Ok(fns) => fns,
+            Err(err) => return err.into_compile_error(),
+        };
+        let regs = self.generate_block_register_definitions(registers, device);
+
+        quote! {
+            #[doc = #module_doc_string]
+            pub mod #snake_case_name {
+                use super::*;
+
+                #doc_attribute
+                #(#cfg_attributes)*
+                pub struct #pascal_case_name<'block, D> {
+                    pub(super) device: &'block mut D,
+                    pub(super) base_address: #address_type,
+                }
+
+                impl<'block, D> #pascal_case_name<'block, D> {
+                    pub fn new(device: &'block mut D, base_address: #address_type) -> Self {
```

Comment:
Do we need this to be pub? Should the user ever use this instead of going through the dedicated generated functions?

Reply:
I could make it `pub(super)`, but I don't know that it really hurts anything to leave it `pub`.

Reply:
Not very important. Let's leave it as it is then